### PR TITLE
Cow Checklist C++ Solution

### PR DESCRIPTION
**Pull request by maggie-j-liu**
I added a C++ solution to Cow Checklist (December 2016 Gold).

### CODE REVIEW

File: solutions/USACO dec16-cow-checklist.mdx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	return pow(a.first - b.first, 2) + pow(a.second - b.second, 2);
+}
+
+int main() 
+{
+	freopen("checklist.in", "r", stdin);
+	freopen("checklist.out", "w", stdout);
+	int h, g;
+	cin >> h >> g;
+	pair<int, int> hcows[h], gcows[g];
```
Comment (bqi343) [inline solutions/USACO dec16-cow-checklist.mdx:33]: As noted [here](https://usaco.guide/bronze/intro-ds?lang=cpp#dynamic-arrays) variable-length arrays aren't part of the C++ standard, so I would prefer using a vector for this purpose.

File: solutions/USACO dec16-cow-checklist.mdx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+```cpp
+#include <iostream>
+#include <cstdio>
+#include <cmath>
+using namespace std;
+
+int energy(pair<int, int> a, pair<int, int> b)
+{
+	return pow(a.first - b.first, 2) + pow(a.second - b.second, 2);
```
Comment (bqi343) [inline solutions/USACO dec16-cow-checklist.mdx:24]: Not relevant for this problem, but `pow` wouldn't be accurate enough for `long long`'s. Ex.

```cpp
int main() {
	const ll x = 2039593823;
	cout << (ll)pow(x,2) << "\n";
	cout << x*x << "\n";
	// 4159942962819755520
	// 4159942962819755329
}
```

Would prefer it if you used your own function to square an integer.

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/USACO dec16-cow-checklist.mdx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	long long dp[2][h + 1][g + 1];
+	for (int i = 0; i <= h; i++)
+	{
+		for (int j = 0; j <= g; j++)
+		{
+			dp[0][i][j] = 1e18;
+			dp[1][i][j] = 1e18;
+		}
+	}
+	dp[0][1][0] = 0;
```
Comment (dongliu0426) [inline solutions/USACO dec16-cow-checklist.mdx:53]: make a note here that farmer john has to start with the first holstein

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

File: solutions/USACO dec16-cow-checklist.mdx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+author: Maggie Liu
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_checklist_gold_dec16.html)
+
+<LanguageSection>
+
+<CPPSection>
+
+**Time Complexity:** $\mathcal{O}(H \cdot G)$
```
[the remark this answers is not in the document]
Comment (maggie-j-liu) [reply solutions/USACO dec16-cow-checklist.mdx:14]: Sure, but [here](https://usaco.guide/general/contributing?lang=cpp#adding-a-solution) the time complexity is inside the CPPSection, so you might want to change that?

[the remark this answers is not in the document]
Comment (bqi343) [reply solutions/USACO dec16-cow-checklist.mdx]: Good point.

File: solutions/USACO dec16-cow-checklist.mdx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	{
+		cin >> hcows[i].first >> hcows[i].second;
+	}
+	for (int i = 0; i < g; i++)
+	{
+		cin >> gcows[i].first >> gcows[i].second;
+	}
+	//dp[0][i][j] is min cost to visit i holsteins and j guernseys ending with holstein
+	//dp[1][i][j] is min cost to visit i holsteins and j guernseys ending with guernsey
+	long long dp[2][h + 1][g + 1];
```
Comment (bqi343) [inline solutions/USACO dec16-cow-checklist.mdx:44]: same comment here as above. either make it constant size or use `vector`s.

```cpp
using ll = long long;
array<vector<vector<ll>>,2> dp;
for (int i = 0; i < 2; ++i) 
  dp[i].assign(h+1,vector<ll>(g+1,1e18));
```

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Review by dongliu0426 — APPROVED** [review]

**Review by bqi343 — APPROVED** [review]

File: solutions/USACO dec16-cow-checklist.mdx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <cstdio>
+using namespace std;
+
+int energy(pair<int, int> a, pair<int, int> b)
+{
+	return (a.first - b.first) * (a.first - b.first) + (a.second - b.second) * (a.second - b.second);
```
Comment (bqi343) [inline solutions/USACO dec16-cow-checklist.mdx:23]: actually some of these lines are > 80 chars

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Pull request merged**